add RNTester-E2E: tests for iOS and Android via Appium, WDIO and Jest (#36267)
facebook/react-native: packages/rn-tester-e2e/README.md

# RNTester E2E folder

In this folder we have a the setup for running E2E testing in RNTester via the usage of [Appium](https://appium.io/) and [WebDriverIO](https://webdriver.io/) and [Jest](https://jestjs.io/).

- [Setting up locally](#setting-up-locally)
  - [(one-off) Setting up Appium](#one-off-setting-up-appium)
  - [Building RNTester app](#building-rntester-app)
    - [Building for iOS](#building-for-ios)
    - [Building for Android](#building-for-android)
  - [Setting up the RNTester E2E folder](#setting-up-the-rntester-e2e-folder)
- [Testing the RNTester app E2E](#testing-the-rntester-app-e2e)
- [Adding new tests (and project structure)](#adding-new-tests-and-project-structure)

## Setting up locally

### (one-off) Setting up Appium

The first step you need to do is to ensure to install the tooling:

```bash

> npm install appium@2.0.0-rc.4 -g

Reviewer: ```suggestion
npm install appium@2.0.0 -g
```

appium driver install uiautomator2
appium driver install xcuitest
```

> More details about drivers in Appium [here](https://appium.github.io/appium/docs/en/2.0/guides/managing-exts/) and [here](https://appium.github.io/appium/docs/en/2.0/quickstart/uiauto2-driver/)

You should not need to run install commands for drivers separately more than once, even if you bump the dep in package.json.

### Building RNTester app

Building manually *.app* and *.apk* is required to run automation tests on local environment.

0. *(optional)* If you previously built RNTester, you may need to clean up build files and Pods:

    ```bash
    yarn test-e2e-local-clean && yarn install
    ```

1. Step 1: install packages for the repository, then navigate in the rn-tester folder

    ```bash
    cd react-native
    yarn install
    cd packages/rn-tester
    ```

> Now, depending on the platform, there are some specific steps

Reviewer 2: can we have a js script that we can invoke with like:
```bash
yarn run-e2e-test <platform> # <platform> can be "ios" or "android"
```

That automate the steps required to build and run the tests?
This will make it easier also to ask people to contribute as they can just write their tests and run the script.

Author: can we have this being a feature to implement as a separate follow up PR? It doesn't sound like a deal breaker and I'd rather get this in and then have smaller PRs after to build on top of it

Reviewer 3: I can handle creating additional PR adding this script, after merging this PR 👍

# RNTester E2E folder

In this folder we have a local setup for doing E2E testing over RNTester via the usage of [Appium](https://appium.io/) and [WebDriverIO](https://webdriver.io/) and [Jest](https://jestjs.io/).

- [Setting up](#setting-up)
  - [(one-off) Setting up Appium](#one-off-setting-up-appium)
  - [Building RNTester app](#building-rntester-app)
    - [Building for iOS](#building-for-ios)
    - [Building for Android](#building-for-android)
  - [Setting up the RNTester E2E folder](#setting-up-the-rntester-e2e-folder)
- [Testing the RNTester app E2E](#testing-the-rntester-app-e2e)
- [Adding new tests (and project structure)](#adding-new-tests-and-project-structure)

## Setting up

### (one-off) Setting up Appium

The first step you need to do is to ensure to install the tooling:

```bash

> npm install appium@2.0.0-beta.71 -g

Reviewer 4: Why do we need to do a global install of this dependency, version using the version provided as a dependency?

Reviewer 5: @NickGerleman Yep it's needed to run the instance of appium server.

appium driver install uiautomator2
appium driver install xcuitest
```

> More details about drivers in Appium [here](https://appium.github.io/appium/docs/en/2.0/guides/managing-exts/) and [here](https://appium.github.io/appium/docs/en/2.0/quickstart/uiauto2-driver/)

You should not need to run install commands for drivers separately more than once, even if you bump the dep in package.json.

### Building RNTester app

Building manually *.app* and *.apk* is required to run automation tests on local environment.

0. *(optional)* If you previously built RNTester, you may need to clean up build files and Pods:

    ```bash

>     yarn test-e2e-local-clean

Reviewer 2: ```suggestion
    yarn test-e2e-local-clean && yarn
```
Just to make sure that the dependencies are reinstalled! ;)

Reviewer 6: Hi @kelset,
This is fixed in the last MR. Could you please resolve this comment?

    ```

1. Step 1: install node modules for the repository, then navigate in the rn-tester folder

    ```bash
    cd react-native
    yarn install
    cd packages/rn-tester
    ```

Now, depending on the platform, there are some specific steps

#### Building for iOS

0. Make sure you have Bundler `gem install bundler` - we use it ensure installing the right version of CocoaPods locally.
1. Install Bundler and CocoaPods dependencies: `bundle install` then `bundle exec pod install` or `yarn setup-ios-hermes` for RNTester with Hermes. In order to use JSC instead of Hermes engine, run: `USE_HERMES=0 bundle exec pod install` or `setup-ios-jsc` instead.
2. Open the generated `RNTesterPods.xcworkspace`.
3. Build the app via XCode.

> 4. Find the **RNTester.app** in `~/Library/Developer/Xcode/DerivedData/RNTesterPods-{id}/Build/Products/Debug-iphonesimulator`

Reviewer 2: I think that we can explore how to tell `xcodebuild` to use a specific derivedData folder. At that point, we can automate these steps.

Reviewer 3: There's `-derivedDataPath <path>` option for that (we're using this one internally in React Native CLI).

Reviewer 2: yeah, I was referring to that. With that parameter, we can automate all these steps and move the build artifact in the right folder, simplifying the setup (and the CI, probably)

Reviewer 6: Hi @kelset,
This is fixed in the last MR. Could you please resolve this comment?

Reviewer 2: What do you mean: `This is fixed in the last MR`? I can still see that the instruction do not explain how to build from command line, passing the `--derivedDataPath`, so it does not seem fixed.

Reviewer 3: Yeah, sorry for miscommunication. I added instructions to `README.md` how to build with RN CLI using `--buildFolder` flag (which corresponds to `-derivedDatpath`). Btw. I used this in CI, and it works great - saved a lot of time :D

5. Copy the app to the following directory `/react-native/packages/rn-tester-e2e/apps`.
6. change its name name to: `rn-tester.app`

#### Building for Android

0. You'll need to have all the [prerequisites](https://reactnative.dev/contributing/how-to-build-from-source#prerequisites) (SDK, NDK) for Building React Native installed.
1. Start an Android emulator.
2. Build the app via

    ```bash
    # In order to not use Hermes engine, run `yarn install-android-jsc` instead.
    yarn install-android-hermes
    yarn start
    ```

    *Note: Building for the first time can take a while.*

3. Find the **RNTester.app** in `~/react-native/packages/rn-tester/android/app/build/outputs/apk/hermes/debug`

> 4. copy the app `app-*-x86_64-debug.apk` to the following directory `/react-native/packages/rn-tester-e2e/apps`

Reviewer 3: I think we shouldn't use here hardcoded architecture `x86_64`, because this may be different for users based on their setup.

Author: ```suggestion
4. copy the app `app-*-debug.apk` to the following directory `/react-native/packages/rn-tester-e2e/apps`
```

5. change its name name to: `rn-tester.apk`

### Setting up the RNTester E2E folder

In `react-native/packages/rn-tester-e2e` install the needed dependencies via:

```bash

> yarn install

Reviewer 2: Out of ignorance, but if the e2e folder is in a workspace, shouldn't `yarn` run in the `react-native` root folder also install all the dependencies in the `rn-tester-e2e` folder? 🤔

```

We need to do this step because the folder is not part of the yarn workspace by design.

Then open the following file

```bash
/react-native/packages/rn-tester-e2e/e2e-config.js
```

> And modify lines L24->L39 to reflect your local setup configuration (ex. `platformVersion`, `deviceName`).

Reviewer 2: Ideally, we should find a way to run tests on both platforms at the same time... 🤔

Reviewer 6: Hi @cipolleschi @kelset,

We checked that implementing this approach will take some time. We want to create the next PR with this as the next step.

> And modify lines L14->L29 to reflect your local setup configuration (ex. `platformVersion`, `deviceName`).

Reviewer 3: ```suggestion
And modify lines L24->L39 to reflect your local setup configuration (ex. `platformVersion`, `deviceName`).
```
